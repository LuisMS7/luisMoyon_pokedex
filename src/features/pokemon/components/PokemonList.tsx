import React from "react";
import { FlatList, ListRenderItemInfo, StyleSheet, View } from "react-native";
import PokemonCard from "@pokemon/components/PokemonCard";
import { Pokemon } from "@pokemon/types/Pokemon.type";
import { perfectSize } from "@utils/perfect-size";

type PokemonListProps = {
	pokemons: Pokemon[];
};

const PokemonList = ({ pokemons }: PokemonListProps) => {
	const renderPokemonCard = (listItem: ListRenderItemInfo<Pokemon>) => {
		const pokemon: Pokemon = listItem.item;
		return <PokemonCard pokemon={pokemon} key={pokemon.name} />;
	};

	const ItemSeparator = () => {
		return <View style={styles.itemSeparator} />;
	};

	return (
		<View style={styles.container}>
			<FlatList
				data={pokemons}
				renderItem={renderPokemonCard}
				horizontal
				ListHeaderComponent={ItemSeparator}
				ItemSeparatorComponent={ItemSeparator}
				ListFooterComponent={ItemSeparator}
			/>
		</View>
	);
};

const styles = StyleSheet.create({
	itemSeparator: {
		width: perfectSize(20),
	},
	container: {
		height: "50%",
	},
});

export default PokemonList;
