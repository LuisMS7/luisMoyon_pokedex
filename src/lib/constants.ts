import { perfectSize } from "@utils/perfect-size";

export const TEXT_SMALL_SIZE = perfectSize(12);

export const TEXT_NORMAL_SIZE = perfectSize(14);

export const TEXT_BIG_SIZE = perfectSize(16);

export const TEXT_SUBTITLE_SIZE = perfectSize(20);

export const TEXT_TITLE_SIZE = perfectSize(24);

export const TEXT_SUPER_SIZE = perfectSize(32);

export const NORMAL_WEIGHT = "400";

export const BOLD_WEIGHT = "800";

export enum API_ENDPOINTS {
	all_pokemons = "api/v2/pokemon",
	pokemons_by_name = "api/v2/pokemon/{pokemon_name}",
}
