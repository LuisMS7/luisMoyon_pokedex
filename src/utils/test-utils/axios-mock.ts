import client from "@lib/axios-instance";
import MockAdapter from "axios-mock-adapter";

export const mockClient = new MockAdapter(client);

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const mockGetOnSuccess = (endpoint: string, mockResponse: any) => {
	mockClient.onGet(endpoint).reply(200, mockResponse);
};

export const mockGetOnFailure = (endpoint: string, statusCode: number) => {
	mockClient.onGet(endpoint).reply(statusCode);
};
