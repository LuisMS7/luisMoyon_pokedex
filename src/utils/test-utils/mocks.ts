import { Pokemon } from "@pokemon/types/pokemon.type";

export const mockPokemon: Pokemon = {
	id: 1,
	name: "mockPokemonName",
	// eslint-disable-next-line max-len
	img: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/1.png",
	types: [{ type: { name: "normal", url: "" }, slot: 1 }],
	weight: 10,
	sprites: {
		back_default:
			// eslint-disable-next-line max-len
			"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/1.png",
		back_female: "",
		back_shiny: "",
		back_shiny_female: "",
		front_default: "",
		front_female: "",
		front_shiny: "",
		front_shiny_female: "",
	},
	moves: [{ move: { name: "normal", url: "" }, version_group_details: [] }],
};
