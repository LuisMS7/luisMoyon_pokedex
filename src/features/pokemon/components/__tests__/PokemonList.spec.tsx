import "react-native";
import React from "react";
import { render, screen } from "@testing-library/react-native";
import PokemonList from "@pokemon/components/PokemonList";
// eslint-disable-next-line max-len
import { PokemonListItemApiResponse } from "@pokemon/types/pokemon-list-api-response.type";
import { mockGetOnSuccess } from "@utils/test-utils/axios-mock";
import { API_ENDPOINTS } from "@lib/constants";

describe("PokemonCard", () => {
	it("should render correctly", async () => {
		const mockPokemons: PokemonListItemApiResponse[] = [
			{
				name: "mockPokemon1",
				url: "",
			},
			{
				name: "mockPokemon2",
				url: "",
			},
			{
				name: "mockPokemon3",
				url: "",
			},
		];
		const firstMockPokemon = {
			id: 1,
			name: "mockPokemon1",
			sprites: {
				other: {
					dream_world: {
						front_default:
							// eslint-disable-next-line max-len
							"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg",
					},
				},
			},
			types: [
				{
					type: {
						name: "grass",
						url: "https://pokeapi.co/api/v2/type/12/",
					},
				},
			],
			weight: 0,
			moves: [],
		};
		const secondMockPokemon = {
			id: 2,
			name: "mockPokemon2",
			sprites: {
				other: {
					dream_world: {
						front_default:
							// eslint-disable-next-line max-len
							"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg",
					},
				},
			},
			types: [
				{
					type: {
						name: "water",
						url: "https://pokeapi.co/api/v2/type/12/",
					},
				},
			],
			weight: 0,
			moves: [],
		};
		const thirdMockPokemon = {
			id: 3,
			name: "mockPokemon3",
			sprites: {
				other: {
					dream_world: {
						front_default:
							// eslint-disable-next-line max-len
							"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg",
					},
				},
			},
			types: [
				{
					type: {
						name: "fire",
						url: "https://pokeapi.co/api/v2/type/12/",
					},
				},
			],
			weight: 0,
			moves: [],
		};
		mockGetOnSuccess(
			API_ENDPOINTS.pokemons_by_name.replace(
				"{pokemon_name}",
				"mockPokemon1",
			),
			firstMockPokemon,
		);
		mockGetOnSuccess(
			API_ENDPOINTS.pokemons_by_name.replace(
				"{pokemon_name}",
				"mockPokemon2",
			),
			secondMockPokemon,
		);
		mockGetOnSuccess(
			API_ENDPOINTS.pokemons_by_name.replace(
				"{pokemon_name}",
				"mockPokemon3",
			),
			thirdMockPokemon,
		);
		const { rerender } = render(
			<PokemonList pokemonsItems={mockPokemons} />,
		);
		await Promise.resolve();
		await rerender(<PokemonList pokemonsItems={mockPokemons} />);
		expect(screen.toJSON()).toMatchSnapshot();
	});
});
