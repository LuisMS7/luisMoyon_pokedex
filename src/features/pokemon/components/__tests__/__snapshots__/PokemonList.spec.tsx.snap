// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonCard should render correctly 1`] = `
<View
  style={
    {
      "height": "50%",
    }
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListFooterComponent={[Function]}
    ListHeaderComponent={[Function]}
    data={
      [
        {
          "id": 1,
          "img": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg",
          "name": "bulbasaur",
          "types": [
            {
              "slot": 1,
              "type": {
                "name": "poison",
                "url": "https://pokeapi.co/api/v2/type/12/",
              },
            },
            {
              "slot": 2,
              "type": {
                "name": "poison",
                "url": "https://pokeapi.co/api/v2/type/4/",
              },
            },
          ],
        },
        {
          "id": 2,
          "img": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/4.svg",
          "name": "bulbasaur2",
          "types": [
            {
              "slot": 1,
              "type": {
                "name": "grass",
                "url": "https://pokeapi.co/api/v2/type/12/",
              },
            },
            {
              "slot": 2,
              "type": {
                "name": "poison",
                "url": "https://pokeapi.co/api/v2/type/4/",
              },
            },
          ],
        },
        {
          "id": 3,
          "img": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/33.svg",
          "name": "bulbasaur3",
          "types": [
            {
              "slot": 1,
              "type": {
                "name": "ghost",
                "url": "https://pokeapi.co/api/v2/type/12/",
              },
            },
            {
              "slot": 2,
              "type": {
                "name": "poison",
                "url": "https://pokeapi.co/api/v2/type/4/",
              },
            },
          ],
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            {
              "width": 29.819916046431132,
            }
          }
        />
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={
          [
            {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            {
              "borderColor": "#A43E9E",
              "borderRadius": 23.855932837144906,
              "borderWidth": 5.963983209286226,
              "flexDirection": "column",
              "height": "100%",
              "width": 521.8485308125448,
            }
          }
        >
          <Text
            style={
              {
                "color": "#A43E9E",
                "fontSize": 29.819916046431132,
                "padding": 29.819916046431132,
                "textAlign": "right",
                "width": "100%",
              }
            }
          >
            #
            0001
          </Text>
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
              }
            }
          />
          <Text
            style={
              {
                "backgroundColor": "#A43E9E",
                "color": "white",
                "fontSize": 35.78389925571736,
                "paddingVertical": 29.819916046431132,
                "textAlign": "center",
              }
            }
          >
            Bulbasaur
          </Text>
        </View>
        <View
          style={
            {
              "width": 29.819916046431132,
            }
          }
        />
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={
          [
            {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            {
              "borderColor": "#74CB48",
              "borderRadius": 23.855932837144906,
              "borderWidth": 5.963983209286226,
              "flexDirection": "column",
              "height": "100%",
              "width": 521.8485308125448,
            }
          }
        >
          <Text
            style={
              {
                "color": "#74CB48",
                "fontSize": 29.819916046431132,
                "padding": 29.819916046431132,
                "textAlign": "right",
                "width": "100%",
              }
            }
          >
            #
            0002
          </Text>
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
              }
            }
          />
          <Text
            style={
              {
                "backgroundColor": "#74CB48",
                "color": "white",
                "fontSize": 35.78389925571736,
                "paddingVertical": 29.819916046431132,
                "textAlign": "center",
              }
            }
          >
            Bulbasaur2
          </Text>
        </View>
        <View
          style={
            {
              "width": 29.819916046431132,
            }
          }
        />
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={
          [
            {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            {
              "borderColor": "#70559B",
              "borderRadius": 23.855932837144906,
              "borderWidth": 5.963983209286226,
              "flexDirection": "column",
              "height": "100%",
              "width": 521.8485308125448,
            }
          }
        >
          <Text
            style={
              {
                "color": "#70559B",
                "fontSize": 29.819916046431132,
                "padding": 29.819916046431132,
                "textAlign": "right",
                "width": "100%",
              }
            }
          >
            #
            0003
          </Text>
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
              }
            }
          />
          <Text
            style={
              {
                "backgroundColor": "#70559B",
                "color": "white",
                "fontSize": 35.78389925571736,
                "paddingVertical": 29.819916046431132,
                "textAlign": "center",
              }
            }
          >
            Bulbasaur3
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            {
              "width": 29.819916046431132,
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
